src/Gui/Tree.cpp

```diff
@@ -3830,22 +3982,33 @@ int DocumentItem::findRootIndex(App::DocumentObject* childObj) {
     if (!TreeParams::getKeepRootOrder() || !childObj || !childObj->isAttachedToDocument())
         return -1;
 
-    // object id is monotonically increasing, so use this as a hint to insert
-    // object back so that we can have a stable order in root level.
+    // Use view provider's tree rank to find correct place at the root level.
 
     int count = this->childCount();
     if (!count)
         return -1;
 
     int first, last;
 
+    auto getTreeRank = [](Gui::ViewProviderDocumentObject* vp) -> int {
+        if (vp->TreeRank.getValue() == -1) {
+            vp->TreeRank.setValue(vp->getObject()->getID());
+            Base::Console().Warning("init treerank to %d\n", vp->getObject()->getID());

```

Comment:
This should at most be log level IMO

Reply:
Oups totally, sorry this is a debug leftover. It can be removed.

Reply:
https://github.com/FreeCAD/FreeCAD/pull/12626